synapse/handlers/sync.py

```diff
@@ -1975,7 +1975,9 @@ async def _get_room_changes_for_incremental_sync(
             # we do send down the room, and with full state, where necessary
 
             old_state_ids = None
-            if room_id in sync_result_builder.joined_room_ids and non_joins:
+            if room_id in sync_result_builder.joined_room_ids and (
+                non_joins or room_id in sync_result_builder.retransmit_room_ids
+            ):
```

Comment:
MV: I think this doesn't change anything---we won't end up using the fact that `room_id` is now `in newly_joined_rooms`.

Reply:
Erik: newly_joined_rooms is a return value, it gets used by other machinery e.g. device_lists. So this is important

Might be easier to initialise `newly_joined_rooms` to a copy of `retransmit_room_ids`

Reply:
0bd54de6401fce068a1697c1fa3e8c36a4025875